File: libraries/ary/ary_sect.f
      SUBROUTINE ARY_SECT( IARY1, NDIM, LBND, UBND, IARY2, STATUS )
*+
*  Name:
*     ARY_SECT

*  Purpose:
*     Create an array section.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL ARY_SECT( IARY1, NDIM, LBND, UBND, IARY2, STATUS )

*  Description:
*     The routine creates a new array section which refers to a
*     selected region of an existing array (or array section). The
*     section may be larger or smaller in extent than the original
*     array.

*  Arguments:
*     IARY = INTEGER (Given)
*        Identifier for the initial array.
*     NDIM = INTEGER (Given)
*        Number of dimensions for new section.
*     LBND( NDIM ) = INTEGER (Given)
*        Lower pixel-index bounds for the new section.
*     UBND( NDIM ) = INTEGER (Given)
*        Upper pixel-index bounds for the new section.
*     IARY2 = INTEGER (Returned)
*        Identifier for the new section.
*     STATUS = INTEGER (Given and Returned)
*        The global status.

*  Notes:
*     -  The number of section dimensions need not match the number of
*     dimensions in the initial array. Pixel-index bounds will be
*     padded with 1's as necessary to identify the pixels to which the
*     new section should refer.
*     -  Note that sections which extend beyond the pixel-index bounds
*     of the initial array will be padded with bad pixels.
*     -  If this routine is called with STATUS set, then a value of
*     ARY__NOID will be returned for the IARY2 argument, although no
*     further processing will occur. The same value will also be
*     returned if the routine should fail for any reason. The ARY__NOID
*     constant is defined in the include file ARY_PAR.

*  Algorithm:
*     -  Set an initial value of ARY__NOID for the IARY2 argument before
*     checking the inherited status.
*     -  Check the bounds of the region to be selected for validity.
*     -  Import the initial array identifier.
*     -  Create the new array entry in the ACB.
*     -  Export an identifier for the new array.
*     -  If an error occurred, then set a value of ARY__NOID for the
*     IACB2 argument and report context information.

*  Copyright:
*     Copyright (C) 1989 Science & Engineering Research Council.
*     All Rights Reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*     
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*     
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 59 Temple Place,Suite 330, Boston, MA
*     02111-1307, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     {enter_new_authors_here}

*  History:
*     2-AUG-1989 (RFWS):
*        Original version.
*     15-SEP-1989 (RFWS):
*        Changed LBND and UBND to be assumed size arrays to prevent
*        run-time errors if NDIM is not valid. Also added a check to
*        prevent ARY1_CUT being called under these circumstances.
*     20-OCT-1989 (RFWS):
*        Renamed to ARY_SECT.
*     {enter_further_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-
      
*  Type Definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'DAT_PAR'          ! DAT_ public constants
      INCLUDE 'ARY_PAR'          ! ARY_ public constants

*  Arguments Given:
      INTEGER IARY1
      INTEGER NDIM
      INTEGER LBND( * )
      INTEGER UBND( * )

*  Arguments Returned:
      INTEGER IARY2

*  Status:
      INTEGER STATUS             ! Global status

*  Local Variables:
      INTEGER IACB1              ! Index of initial array in the ACB
      INTEGER IACB2              ! Index of new array in the ACB

*.

*  Set an initial value for the IARY2 argument.
      IARY2 = ARY__NOID

*  Check inherited global status.
      IF ( STATUS .NE. SAI__OK ) RETURN

*  Import the initial array identifier.
      CALL ARY1_IMPID( IARY1, IACB1, STATUS )

*  Check the bounds of the region to select for validity.
      CALL ARY1_VBND( NDIM, LBND, UBND, STATUS )

*  Create an ACB entry for the new array.
      IF ( STATUS .EQ. SAI__OK ) THEN
         CALL ARY1_CUT( IACB1, NDIM, LBND, UBND, IACB2, STATUS )
      END IF

*  Export an identifier for the new array.
      CALL ARY1_EXPID( IACB2, IARY2, STATUS )

*  If an error occurred, then set a value of ARY__NOID for the IARY2
*  argument, report context information and call the error tracing
*  routine.
      IF ( STATUS .NE. SAI__OK ) THEN
         IARY2 = ARY__NOID
         CALL ERR_REP( 'ARY_SECT_ERR',
     :   'ARY_SECT: Error creating an array section.', STATUS )
         CALL ARY1_TRACE( 'ARY_SECT', STATUS )
      END IF       

      END

File: libraries/ary/arySect.c
#include "sae_par.h"
#include "ary1.h"
#include "mers.h"
#include "dat_par.h"

void arySect( Ary *ary1, int ndim, const hdsdim *lbnd, const hdsdim *ubnd,
              Ary **ary2, int *status ) {
/*
*+
*  Name:
*     arySect

*  Purpose:
*     Create an array section.

*  Synopsis:
*     void arySect( Ary *ary1, int ndim, const hdsdim *lbnd,
*                   const hdsdim *ubnd, Ary **ary2, int *status )

*  Description:
*     This function creates a new array section which refers to a
*     selected region of an existing array (or array section). The
*     section may be larger or smaller in extent than the original
*     array.

*  Parameters:
*     ary
*        Identifier for the initial array.
*     ndim
*        Number of dimensions for new section. This is the length of the
*        "lbnd" and Ubnd" arrays.
*     lbnd
*        Lower pixel-index bounds for the new section.
*     ubnd
*        Upper pixel-index bounds for the new section.
*     ary2
*        Address of a variable in which to return an identifier for the
*        new section.
*     status
*        The global status.

*  Notes:
*     -  If the supplied array is locked read-only by the current thread
*     an attempt will be made to promote the lock to a read-write lock
*     and an error will be reported if this attempt fails. This promotion
*     is necessary because the meta-data associated with the cut (e.g.
*     pixel shift etc) can be modified using the returned identifier.
*     -  The number of section dimensions need not match the number of
*     dimensions in the initial array. Pixel-index bounds will be
*     padded with 1's as necessary to identify the pixels to which the
*     new section should refer.
*     -  Note that sections which extend beyond the pixel-index bounds
*     of the initial array will be padded with bad pixels.
*     -  If this routine is called with "Status" set, then a value of
*     NULL will be returned for the "ary2" argument, although no
*     further processing will occur. The same value will also be
*     returned if the routine should fail for any reason.

*  Copyright:
*      Copyright (C) 2017 East Asian Observatory
*      All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  History:
*     03-JUL-2017 (DSB):
*        Original version, based on equivalent Fortran routine by RFWS.

*-
*/

/* Local variables: */
   AryACB *acb1;                 /* Pointer to initial array in the ACB */
   AryACB *acb2 = NULL;          /* Pointer to new array in the ACB */

/* Set an initial value for the "ary2" argument. */
   *ary2 = NULL;

/* Check inherited global status. */
   if( *status != SAI__OK ) return;

/* Import the initial array identifier. */
   acb1 = (AryACB *) ary1Impid( ary1, 1, 1, 1, status );

/* Check the bounds of the region to select for validity. */
   ary1Vbnd( ndim, lbnd, ubnd, status );

/* Create an ACB entry for the new array. */
   if( *status == SAI__OK ){
      ARY__DCB_LOCK_MUTEX;
      ARY__ACB_LOCK_MUTEX;
      ary1Cut( acb1, ndim, lbnd, ubnd, &acb2, status );
      ARY__ACB_UNLOCK_MUTEX;
      ARY__DCB_UNLOCK_MUTEX;
   }

/* Export an identifier for the new array. */
   *ary2 = ary1Expid( (AryObject *) acb2, status );

/* If an error occurred, then set a value of NULL for the "ary2"
   argument, report context information and call the error tracing
   routine. */
   if( *status != SAI__OK ){
      *ary2 = NULL;
      errRep( " ", "arySect: Error creating an array section.",
              status );
      ary1Trace( "arySect", status );
   }

}
